### PR TITLE
Add request parameters variable

### LINKED ISSUE #18: Add option to have user defined `params` for the OpenAI request
## Why?

To use the [Azure-hosted OpenAI rest API](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference) the `api-version` has to be provided in the query:

```
POST https://YOUR_RESOURCE_NAME.openai.azure.com/openai/deployments/YOUR_DEPLOYMENT_NAME/completions?api-version=2022-12-01
```

At the moment, I can use this library by overriding the `openai-request` method, like so:

``` elisp
  (defmacro openai-request (url &rest body)
    "Wrapper for `request' function.

  The URL is the url for `request' function; then BODY is the arguments for rest."
    (declare (indent 1))
    `(progn
       (setq openai-error nil)
       (request ,url
         :params '(("api-version" . "2023-03-15-preview"))
         :error (cl-function
                 (lambda (&key response &allow-other-keys)
                   (setq openai-error response)
                   (openai--handle-error response)))
         ,@body))))
```

I would like it if the library supports this feature.

## What?

A more generic approach would be:

1. Create a variable

``` elisp
(defcustom openai-request-parameters '()
  "The parameters for the OpenAI request."
  :type 'list
  :group 'openai)
```

(Maybe name the variable `openai-user-defined-request-parameters` to emphasize that the library __might__ add other parameters.)

2. Then pass the parameters to all OpenAI request, for example in the chat:

``` elisp
(cl-defun openai-chat ( messages callback
                        &key
                        (base-url openai-base-url)
                        (params openai-user-defined-params)
                        (content-type "application/json")
                        (key openai-key)
                        org-id
                        (model "gpt-3.5-turbo")
                        temperature
                        top-p
                        n
                        stream
                        stop
                        max-tokens
                        presence-penalty
                        frequency-penalty
                        logit-bias
                        (user openai-user))
  "Send chat request.

Arguments MESSAGES and CALLBACK are required for this type of request.  MESSAGES
is the conversation data.  CALLBACK is the execuation after request is made.

Arguments BASE-URL, PARAMS, CONTENT-TYPE, KEY, ORG-ID and USER are global options; however, you
can overwrite the value by passing it in.

The rest of the arugments are optional, please see OpenAI API reference page
for more information.  Arguments here refer to MODEL,  TEMPERATURE, TOP-P, N,
STREAM, STOP, MAX-TOKENS, PRESENCE-PENALTY, FREQUENCY-PENALTY, and LOGIT-BIAS."
  (openai-request (concat base-url "/chat/completions")
    :type "POST"
    :params params         ;; <--- does this need an `@` or `,` prefix?
    :headers (openai--headers content-type key org-id)
    :data (openai--json-encode
           `(("model"             . ,model)
             ("messages"          . ,messages)
             ("temperature"       . ,temperature)
             ("top-p"             . ,top-p)
             ("n"                 . ,n)
             ("stream"            . ,stream)
             ("stop"              . ,stop)
             ("max_tokens"        . ,max-tokens)
             ("presence_penalty"  . ,presence-penalty)
             ("frequency_penalty" . ,frequency-penalty)
             ("logit_bias"        . ,logit-bias)
             ("user"              . ,user)))
    :parser 'json-read
    :complete (cl-function
               (lambda (&key data &allow-other-keys)
                 (funcall callback data)))))
```

3. I think the macro passes the parameter automagically to the underlying request:

``` elisp
(defmacro openai-request (url &rest body)
  "Wrapper for `request' function.

The URL is the url for `request' function; then BODY is the arguments for rest."
  (declare (indent 1))
  `(progn
     (setq openai-error nil)
     (request ,url
       :error (cl-function
               (lambda (&key response &allow-other-keys)
                 (setq openai-error response)
                 (openai--handle-error response)))
       ,@body)))
```

### PR DESCRIPTION
**Pull request by JCZuurmond**
Resolves #18

### CODE REVIEW

File: openai-fine-tune.el [hunk 1]
```diff
@@ -154,16 +162,17 @@ you can overwrite the value by passing it in."
 (cl-defun openai-fine-tune-list-events ( fine-tune-id callback
                                          &key
                                          (content-type "application/json")
+                                         (base-url openai-base-url)
```
[the remark this answers is not in the document]
Comment (jcs090218) [reply openai-fine-tune.el:165]: Oh, yeah. Can you fix this for me? Thank you!

File: openai-audio.el [hunk 2]
```diff
@@ -49,14 +50,15 @@ Argument FILE is audio file to transcribe, in one of these formats: mp3, mp4,
 mpeg, mpga, m4a, wav, or webm.  CALLBACK is the execuation after request is
 made.
 
-Arguments BASE-URL, CONTENT-TYPE, KEY and ORG-ID are global options; however,
+Arguments BASE-URL, PARAMETERS, CONTENT-TYPE, KEY and ORG-ID are global options; however,
```
Comment (jcs090218) [inline openai-audio.el:53]: I prefer max 80 column per line, can you fix that for me? (And all other places)

File: openai-file.el [hunk 3]
```diff
@@ -88,6 +92,7 @@ you can overwrite the value by passing it in."
 
 (cl-defun openai-file-delete ( file-id callback
                                &key
+                               (parameters openai-parameters)
```
Comment (jcs090218) [inline openai-file.el:95]: I think this should be under `base-url openai-base-url)`.

**Review by jcs090218 — COMMENTED** [review]
Some nits! Otherwise, LGTM! :D

File: openai-audio.el (hunk 2 above)
Comment (JCZuurmond) [reply openai-audio.el]: done

**Comment by JCZuurmond** [comment]
@jcs090218 : Comments are fixed. I tested the chat with these changes, and it works!

File: openai-fine-tune.el [hunk 4]
```diff
@@ -154,17 +162,18 @@ you can overwrite the value by passing it in."
...
+                                         (parameters openai-parameters)
```
[the remark this answers is not in the document]
Comment (JCZuurmond) [reply openai-fine-tune.el:166]: yes!

**Comment by JCZuurmond** [comment]
Again, the comments are fixed!

**Pull request merged**